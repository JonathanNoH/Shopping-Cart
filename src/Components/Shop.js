import React, { useState } from "react";
import Item from "./Item";

const Shop = () => {
  const [cart, setCart] = useState({});

  const addItemToCart = (item, currentAmount) => {
    setCart(prevState => {
      return { ...prevState, [item]: (currentAmount + 1)}
    })
  }

  return (
    <div>
      <h1>Shop</h1>
      <Item 
        title="Apple" 
        description="fruit" 
        amount={('Apple' in cart) ? cart.Apple : 0 }
        onAdd={addItemToCart}
      />
    </div>
  )
}

export default Shop;