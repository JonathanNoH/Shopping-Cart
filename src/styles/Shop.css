h1 {
  color: var(--main-blue);
  grid-column: 1 / -1;
  place-self: center;
}

.Shop {
  position: relative;
}

.container {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));

}