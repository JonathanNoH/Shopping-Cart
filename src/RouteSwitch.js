import React from "react";
import {
  BrowserRouter,
  Routes,
  Route
} from "react-router-dom";

import App from "./App";
import Home from "./Components/Home";
import Shop from "./Components/Shop";

const Router = () => {
  
  return (
    <BrowserRouter basename="https://jonathannoh.github.io/Shopping-Cart">
      <Routes>
        <Route path="/" element={<App />}>
          <Route index element={<Home />}/>
          <Route path="shop" element={<Shop />}/>
        </Route>
      </Routes>
    </BrowserRouter>
  )
}

export default Router;