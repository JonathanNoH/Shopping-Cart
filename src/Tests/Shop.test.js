import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";  // optional
import userEvent from "@testing-library/user-event";
import Shop from "../Components/Shop";
import RouteSwitch from "../RouteSwitch";

// This used to work without rendering everything but I could not figure out how to mock or fix useOutletContext() calls inside Shop when not rendering the Routing


describe('Shop Component', () => {
  it("renders shop", () => {
    
    const { container } = render(<RouteSwitch />);
    const shopButton = screen.getByRole('link', { name: "Shop"});
    userEvent.click(shopButton);
    expect(container).toMatchSnapshot();
  })
});

describe('Shop Logic', () => {
  it("adds item to cart", () => {
    render(<RouteSwitch />);
    const shopButton = screen.getByRole('link', { name: "Shop"});
    userEvent.click(shopButton);

    const button = screen.getByRole('button', { name: "Add Apple To Cart" });
    const amount = screen.getByTestId('Apple-test-amount');
    const amountToAdd = screen.getByRole('spinbutton', { name: "Apples to add:"});
    expect(amount.textContent).toBe("0");

    userEvent.click(button);

    expect(amount.textContent).toBe((0 + parseInt(amountToAdd.value)).toString());
  })
  it("adds multiple items to cart", () => {
    render(<RouteSwitch />);
    const shopButton = screen.getByRole('link', { name: "Shop"});
    userEvent.click(shopButton);
    const button = screen.getByRole('button', { name: "Add Apple To Cart" });
    const amount = screen.getByTestId('Apple-test-amount');
    const amountToAdd = screen.getByRole('spinbutton', { name: "Apples to add:"});
    expect(amount.textContent).toBe("0");

    userEvent.type(amountToAdd, "{backspace}12");
    expect(amountToAdd.value).toBe("12");

    userEvent.click(button);

    expect(amount.textContent).toBe("12");
  })
  it('removes items from cart', () => {
    render(<RouteSwitch />);
    const shopButton = screen.getByRole('link', { name: "Shop"});
    userEvent.click(shopButton);

    const button = screen.getByRole('button', { name: "Add Apple To Cart" });
    const amount = screen.getByTestId('Apple-test-amount');
    const amountToAdd = screen.getByRole('spinbutton', { name: "Apples to add:"});

    expect(amount.textContent).toBe("0");
    userEvent.click(button);
    expect(amount.textContent).toBe((0 + parseInt(amountToAdd.value)).toString());

    const cartButton = screen.getByRole('button', { name: 'Cart'});
    userEvent.click(cartButton);
    const removeButton = screen.getByRole('button', { name: "Remove Apple from cart" })

    userEvent.click(removeButton);
    expect(amount.textContent).toBe("0");
  })
})