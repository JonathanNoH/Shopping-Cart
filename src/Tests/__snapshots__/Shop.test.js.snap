// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop Component renders shop 1`] = `
<div>
  <div>
    <h1>
      Shop
    </h1>
    <div>
      <h3>
        Apple
      </h3>
      <div
        role="Apple-description"
      >
        fruit
      </div>
      <form>
        <label
          for="amount-to-add-Apple"
        >
          Apple
          s to add: 
        </label>
        <input
          id="amount-to-add-Apple"
          max="99"
          min="0"
          name="amount-to-add-Apple"
          type="number"
          value="1"
        />
        <button
          aria-label="Add Apple To Cart"
          type="submit"
        >
          Add To Cart
        </button>
      </form>
      <div>
        <span
          role="Apple-in-cart"
        >
          0
        </span>
         in cart
      </div>
    </div>
  </div>
</div>
`;
